ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java

```diff
@@ -195,25 +195,33 @@ void handleResponse(ResponseInfo responseInfo, DeleteResponse deleteResponse) {
               new RouterException("Received wrong response that is not for the corresponding request.",
                   RouterErrorCode.UnexpectedInternalError));
         } else {
-          ServerErrorCode getError = deleteResponse.getError();
-          if (getError == ServerErrorCode.No_Error || getError == ServerErrorCode.Blob_Deleted) {
+          ServerErrorCode serverError = deleteResponse.getError();
+          if (serverError == ServerErrorCode.No_Error || serverError == ServerErrorCode.Blob_Deleted) {
             operationTracker.onResponse(replica, TrackedRequestFinalState.SUCCESS);
             if (RouterUtils.isRemoteReplica(routerConfig, replica)) {
               logger.trace("Cross colo request successful for remote replica {} in {} ", replica.getDataNodeId(),
                   replica.getDataNodeId().getDatacenterName());
               routerMetrics.crossColoSuccessCount.inc();
             }
+          } else if (serverError == ServerErrorCode.Disk_Unavailable) {
+            logger.trace("Replica {} returned Disk_Unavailable for a delete request with correlationId : {} ", replica,
+                deleteRequest.getCorrelationId());
+            operationTracker.onResponse(replica, TrackedRequestFinalState.DISK_DOWN);
+            setOperationException(
+                new RouterException("Server returned: " + serverError, RouterErrorCode.AmbryUnavailable));
+            routerMetrics.routerRequestErrorCount.inc();
+            routerMetrics.getDataNodeBasedMetrics(replica.getDataNodeId()).deleteRequestErrorCount.inc();
```

Comment:
`processServerError` indeed returns AmbryUnavailable. The reason I put these statements in else-if block is I need to skip `onErrorResponse` in which it sets TrackedRequestFinalState. Also, I cannot change `fromRouterErrorCodeToFinalState` directly because it cannot tell AmbryUnavailable is `Disk_Unavailable` or `Replica_Unavailable`. For latter one, the replica could be temporarily stopped and restarting it will bring intact data (blob still exists in that replica). Hence, I copied statements to this else-if block without changing logic in `fromRouterErrorCodeToFinalState`.

Reply:
we are not restarting the replica in frontend. Since processServerError returns the same RouterErrorCode from ReplicaUnavailable and DiskUnavailable, then we treat them as the same error in the router code. I didn't see in any operation code that we treat these two errors differently.